[HACK Week] Order Navigation

toupper opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
With this PR we implement order navigation within the order details. With this, users can navigate to previous or next orders without the need to go back to the order list and tap on an order again. This was requested a couple of times through AppStore/Google Play reviews.
If the Order Detail screen is presented through a notification or shortcut these buttons are hidden.

While on Android the common pattern for this kind of navigation is scrolling horizontally, we decided that given our current implementation the effort would be too much. Because of that, we replicate the iOS approach and add two buttons to the top bar.

### Known caveats
With this implementation, the user can only visit those orders that were retrieved when loading the order list. That means that if there's only one page loaded the user will be bound to those orders. Even with that and given the size of our pages (25) the users can still benefit from it.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Go to orders
2. Tap on an order
3. See that the open order is the one that was selected previously
4. See that you can navigate to previous and next orders, but when you reach a bound -upper or lower- the matching button is disabled.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-android/assets/1864060/64d0268d-9514-4156-a56c-c6934c1c410a

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

toupper committed:
Add buttons for previous and next order

toupper committed:
WIP: Pass list of order ids

toupper committed:
Add orderIds argument

toupper committed:
Navigate to the next order when the button is clicked

toupper committed:
Enable back button and pass order ids array

toupper committed:
Enable/disable order navigation buttons according to position

toupper committed:
Simplify navigation functions by removing the list position parameter and rely only on orderId and orderIds

toupper committed:
Rename to a more meaningful name.

toupper committed:
Add tests for enabling order navigation

toupper committed:
Hide order navigation buttons when there's only one order.

toupper committed:
Add tests for navigation order buttons visibility

toupper committed:
Fix navigation

toupper committed:
Remove empty line

kidinov commented:
I was just curious to look into it. @toupper great work!

I've just noticed that it crashes if you click fast enough on the next order arrow

```
java.util.concurrent.CompletionException: java.util.NoSuchElementException: No value present
at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java: 308)
at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java: 323)
at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java: 684)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:486)
at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java: 2138)
at com.google.android.apps.nexuslauncher.allapps.H0.onResult(Unknown Source: 59)
at android.app.appsearch.SearchSessionUtil$1.lambda$onResult$0(SearchSessionUtil.java: 129)
at android.app.appsearch.SearchSessionUtil$1.$r8$lambda$_nlk9WiPjSfaAsqKYmp0aPvSRC0(SearchSessionUtil.java: 1)
at android.app.appsearch.SearchSessionUtil$1$$ExternalSyntheticLambda0.run(R8$$SyntheticClass: 13)
at android.os.Handler.handleCallback(Handler.java: 958)
at android.os.Handler.dispatchMessage(Handler.java: 99)
at android.os.Looper.loopOnce(Looper.java: 205)
at android.os.Looper.loop(Looper.java: 294)
at android.os.HandlerThread.run(HandlerThread.java: 67)
```

Also, UX wise it could be better as swipable thing, e.g swipe right shows the next order

rossanafmenezes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListAdapter.kt:
```diff
@@ -83,6 +83,14 @@ class OrderListAdapter(
+            } else {
+                 null
+            }
+        } ?: listOf<Long>()
```
NP: The Long type here is inferred. No need to add it.

toupper replied:
Awesome! Thanks for the tip. Removed in f44b0ca3ba3df3a642c18aa8b401d40801e63095

rossanafmenezes reviewed (commented):
Hi @toupper 👋 

The code looks good. I left one NP comment. 
Also, this could be included in the Release notes, WDYT?

Another thing is that I tried tapping on the arrow (either one)  multiple times (I know this is not a behaviour we expect but... ) and the app crashed:

```
FATAL EXCEPTION: main

                 Process: [[com.woocommerce.android.dev](http://com.woocommerce.android.dev/)](http://com.woocommerce.android.dev), PID: 5168

                 java.lang.IllegalArgumentException: Required value was null.

                 	at [[com.woocommerce.android](http://com.woocommerce.android/)](http://com.woocommerce.android).ui.orders.details.OrderDetailViewModel.getOrder(OrderDetailViewModel.kt:118)

                 	at [[com.woocommerce.android](http://com.woocommerce.android/)](http://com.woocommerce.android).ui.orders.details.OrderDetailViewModel.onNextOrderClicked(OrderDetailViewModel.kt:325)

                 	at [[com.woocommerce.android](http://com.woocommerce.android/)](http://com.woocommerce.android).ui.orders.details.OrderDetailFragment.onMenuItemSelected(OrderDetailFragment.kt:221)

                 	at androidx.core.view.MenuHostHelper.onMenuItemSelected([[MenuHostHelper.java:107](http://menuhosthelper.java:107/)](http://MenuHostHelper.java:107))

                 	at androidx.activity.ComponentActivity.onMenuItemSelected([[ComponentActivity.java:518](http://componentactivity.java:518/)](http://ComponentActivity.java:518))

                 	at [[androidx.fragment.app](http://androidx.fragment.app/)](http://androidx.fragment.app).FragmentActivity.onMenuItemSelected([[FragmentActivity.java:265](http://fragmentactivity.java:265/)](http://FragmentActivity.java:265))

                 	at [[androidx.appcompat.app](http://androidx.appcompat.app/)](http://androidx.appcompat.app).AppCompatActivity.onMenuItemSelected([[AppCompatActivity.java:264](http://appcompatactivity.java:264/)](http://AppCompatActivity.java:264))

                 	at androidx.appcompat.view.WindowCallbackWrapper.onMenuItemSelected([[WindowCallbackWrapper.java:109](http://windowcallbackwrapper.java:109/)](http://WindowCallbackWrapper.java:109))

                 	at [[androidx.appcompat.app](http://androidx.appcompat.app/)](http://androidx.appcompat.app).ToolbarActionBar$2.onMenuItemClick([[ToolbarActionBar.java:66](http://toolbaractionbar.java:66/)](http://ToolbarActionBar.java:66))

                 	at androidx.appcompat.widget.Toolbar$1.onMenuItemClick([[Toolbar.java:221](http://toolbar.java:221/)](http://Toolbar.java:221))

                 	at androidx.appcompat.widget.ActionMenuView$MenuBuilderCallback.onMenuItemSelected([[ActionMenuView.java:781](http://actionmenuview.java:781/)](http://ActionMenuView.java:781))

                 	at [[androidx.appcompat.view.menu](http://androidx.appcompat.view.menu/)](http://androidx.appcompat.view.menu).MenuBuilder.dispatchMenuItemSelected([[MenuBuilder.java:834](http://menubuilder.java:834/)](http://MenuBuilder.java:834))

                 	at [[androidx.appcompat.view.menu](http://androidx.appcompat.view.menu/)](http://androidx.appcompat.view.menu).MenuItemImpl.invoke([[MenuItemImpl.java:158](http://menuitemimpl.java:158/)](http://MenuItemImpl.java:158))

                 	at [[androidx.appcompat.view.menu](http://androidx.appcompat.view.menu/)](http://androidx.appcompat.view.menu).MenuBuilder.performItemAction([[MenuBuilder.java:985](http://menubuilder.java:985/)](http://MenuBuilder.java:985))

                 	at [[androidx.appcompat.view.menu](http://androidx.appcompat.view.menu/)](http://androidx.appcompat.view.menu).MenuBuilder.performItemAction([[MenuBuilder.java:975](http://menubuilder.java:975/)](http://MenuBuilder.java:975))

                 	at androidx.appcompat.widget.ActionMenuView.invokeItem([[ActionMenuView.java:625](http://actionmenuview.java:625/)](http://ActionMenuView.java:625))

                 	at [[androidx.appcompat.view.menu](http://androidx.appcompat.view.menu/)](http://androidx.appcompat.view.menu).ActionMenuItemView.onClick([[ActionMenuItemView.java:151](http://actionmenuitemview.java:151/)](http://ActionMenuItemView.java:151))

                 	at android.view.View.performClick([[View.java:7506](http://view.java:7506/)](http://View.java:7506))

                 	at android.view.View.performClickInternal([[View.java:7483](http://view.java:7483/)](http://View.java:7483))

                 	at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)

                 	at android.view.View$[[PerformClick.run](http://performclick.run/)](http://PerformClick.run)([[View.java:29334](http://view.java:29334/)](http://View.java:29334))

                 	at android.os.Handler.handleCallback([[Handler.java:942](http://handler.java:942/)](http://Handler.java:942))

                 	at android.os.Handler.dispatchMessage([[Handler.java:99](http://handler.java:99/)](http://Handler.java:99))

                 	at android.os.Looper.loopOnce([[Looper.java:201](http://looper.java:201/)](http://Looper.java:201))

                 	at android.os.Looper.loop([[Looper.java:288](http://looper.java:288/)](http://Looper.java:288))

                 	at [[android.app](http://android.app/)](http://android.app).ActivityThread.main([[ActivityThread.java:7872](http://activitythread.java:7872/)](http://ActivityThread.java:7872))

                 	at java.lang.reflect.Method.invoke(Native Method)

                 	at [[com.android](http://com.android/)](http://com.android).internal.os.RuntimeInit$[[MethodAndArgsCaller.run](http://methodandargscaller.run/)](http://MethodAndArgsCaller.run)([[RuntimeInit.java:548](http://runtimeinit.java:548/)](http://RuntimeInit.java:548))

                 	at [[com.android](http://com.android/)](http://com.android).internal.os.ZygoteInit.main([[ZygoteInit.java:936](http://zygoteinit.java:936/)](http://ZygoteInit.java:936))

08:25:27.389  I  Uncaught exception received.
```

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/OrderDetailViewModelTest.kt:
```diff
@@ -1907,4 +1934,77 @@ class OrderDetailViewModelTest : BaseUnitTest() {
     }
+
+    @Test
+    fun `when the order is the first one then it disables navigation to previous order`() = testBlocking {
```
That could be something like 

`given order ids passed, when ViewModel starts, then previous order navigations is not enabled`

(pay attention to the commas - https://github.com/woocommerce/woocommerce-android/blob/trunk/CONVENTION.md#testing)

toupper replied:
Thanks for the suggestion and the link! Done in 19d298cb9ca39d25c665c91f7cc48df5be822032

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderNavigator.kt:
```diff
@@ -187,6 +189,23 @@ class OrderNavigator @Inject constructor() {
+                        target.allOrderIds)
+                    .let {
+                        // Replace the current order detail without animating the transition
+                        val navOptions = NavOptions.Builder()
```
I think you can set this up in the action itself (in XML), and there you can use `app:enterAnim="@null"`, so you can avoid  creating `nav_enter_anim` file

toupper replied:
Nice! That's much easier and works, done in 4bb8be33eddb77e1bbda313efe790871e7505cd6

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListAdapter.kt:
```diff
@@ -83,6 +83,14 @@ class OrderListAdapter(
 
     override fun onBindViewHolder(holder: ViewHolder, position: Int) {
         val item = getItem(position)
+        val allOrderIds = getCurrentList()?.toList()?.mapNotNull {
```
`onBindViewHolder` is called every time the user scrolls the list, and here we take the list, convert it to a new list, and then iterate over all the items in the list. This list can be huge, and all that is on the main thread. (too many "lists" in one sentence  😀)

You may consider overriding `submit list` in here and building the list of the orderids you want to pass once per list submission.

hichamboushaba replied:
If I may add something else, related to this point, it seems we are passing the list of all of the orders as a navigation argument, this could lead to throwing a [`TransactionTooLargeException`](https://developer.android.com/reference/android/os/TransactionTooLargeException) for big stores with a large number of orders. I don't think the list of `ids` would cause this alone, but when considering the rest of the data that we might be already saving in the Saved State's bundle, it could add up, and we end up taking more than the allowed memory. Can't we move the logic of getting the list of orders to the Order Details screen?

toupper replied:
Thanks for your comment @hichamboushaba, that information is very valuable for me since I'm starting with Android development.
On that link, we can find that we have 1MB of buffer shared by all transactions in progress for the process. If we take into account that an order id is an instance of a long type (8 bytes of size), we would need something in the ballpark of 131072 ids to fill that buffer, a huge number even for stores with many orders. Not only that, the user would have to load all these orders on screen via pagination, which seems even more unlikely.
Because of that, we (@malinajirka and I) have the opinion that it's better to leave it as it is and assume that risk given the low chance of it happening, keeping the simplicity of passing the ids and not having to retrieve them again in the view model. (Retrieving the ids in the view model would be complex as we would have to pass as well what filters were enabled in the order list to keep the navigation identical).

hichamboushaba replied:
Thanks for looking into this @toupper 🙏, and it was not a blocking point, so if you and @malinajirka think it's OK, then go ahead and keep it.

But I just want to clarify some points:
> we can find that we have 1MB of buffer shared by all transactions in progress for the process

Even though the documentation says 1MB, the reality is that the amount of memory depends on the device, some devices have only 512KB and some have 1MB, and depending on OEMs, there could be devices with other values, just as an example, if you check some example of the [exception](https://github.com/woocommerce/woocommerce-android/issues/2594) that's thrown for our app, you can see that we get the exception for less way less then 1MB.

> If we take into account that an order id is an instance of a long type (8 bytes of size), we would need something in the ballpark of 131072 ids to fill that buffer

I know 🙂, and that's why I said above that the data would add up, I didn't mean that we'll have 1MB or even 500KB of the order's IDs, I meant that passing a slightly larger data as a navigation argument would increase the chances getting more of these exceptions, because as you said, the memory is shared, and all of the activities/fragments in backstack with their navigation arguments and saved states contribute to it.

I raised this point because this is something we were careless about in the past 🤷‍♂️, and it's one of the long-lasting issues in our backlog, and which is not easy to fix when the code was written in a certain way, so I just wanted to be sure we are aware of it, but as said above, it's not a blocking point.

toupper committed:
Remove explicit type.

toupper committed:
Rename following test naming convention

toupper committed:
Set null animation in the  nav graph orders xml instead of having to create a new one

toupper committed:
Use nav args instead of the order property because the latter is nil at times

toupper commented:
@kidinov @rossanafmenezes Thanks for testing and finding the crash! I added a fix for that in 1bd1d7cbc7ed0cf8dff17faf1ef7e766caa46e10. Basically, I was using the order computer property in OrderDetailViewModel that relied on `viewState.orderInfo` not being null (`requireNotNull`) but unfortunately it was null at times. Now I use the nav args and I couldn't reproduce the crash again.

I see that both your stack traces are different, with my fix I addressed @rossanafmenezes 's one. @kidinov Do you think the fix I added also addresses the crash you found?

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderNavigator.kt:
```diff
@@ -187,6 +189,21 @@ class OrderNavigator @Inject constructor() {
+                        target.orderId,
+                        target.allOrderIds)
+                    .let {
+                        // Replace the current order detail without animating the transition
```
np: the comment is outdated now

Also, I forgot to mention that `setPopUpTo(R.id.orders, false)` is the same as you set already `app:popUpTo="@+id/orders"`. You probably need to add `app:popUpToInclusive="false"` though

toupper replied:
True, done in 480b2a9990a60b0dfdbe4407d56e70f548b1421e

toupper committed:
Create the allOrderIds in submitList instead of onBindViewHolder to avoid calling it multiple times.

toupper committed:
Move pop up logic to xml

toupper committed:
Fix tests by passing missing argument

rossanafmenezes commented:
> @kidinov @rossanafmenezes Thanks for testing and finding the crash! I added a fix for that in https://github.com/woocommerce/woocommerce-android/commit/1bd1d7cbc7ed0cf8dff17faf1ef7e766caa46e10. Basically, I was using the order computer property in OrderDetailViewModel that relied on viewState.orderInfo not being null (requireNotNull) but unfortunately it was null at times. Now I use the nav args and I couldn't reproduce the crash again.

👋  @toupper 

Thanks for fixing it. It looks like the crash has been fixed. That said I noticed something else that I believe was not happening before which is the behaviour of the menu bar when you click on the arrows:

<img src="https://github.com/woocommerce/woocommerce-android/assets/30724184/07a32b78-f7d5-4cb4-ae12-89472b86350c" width =350>

It's blinking and the whole edit item is being dragged to the left while the order number is loaded 🤔 


Also, I'm not sure if this is the same as in iOS, so just in case it isn't, regarding the failing CI Test `buildkite/woocommerce-android/pipeline-pipeline-upload — Failed (exit status 1)` I asked here: p1696951226217509-slack-C02QANACA

You can simply retry the test here and that should solve the issue: https://buildkite.com/automattic/woocommerce-android/builds/15690#018b14c7-e9bf-4536-a26c-4f65650f478d

toupper committed:
Making allOrderIds required made the app crash when it was not necessary. Here we make it nullable to fix that.

toupper committed:
Fix detekt issues

toupper committed:
Rename ViewState to OrderDetailViewState

toupper committed:
Move classes to their own file to prevent class too large detekt issue

malinajirka committed:
Extract tracking from OrderDetailViewModel

malinajirka committed:
Update tests after extraction of tracking

malinajirka committed:
Fix detekt

toupper committed:
Merge pull request #9965 from woocommerce/hack-week/order-navigation-extract-tracking

Extract Tracking From OrderDetailViewModel

toupper committed:
Set buttons visibility to false

toupper committed:
Merge branch 'trunk' into hack-week/order-navigation

toupper committed:
Add release note

toupper commented:
Thanks for the review and comments @rossanafmenezes! I fixed the detekt CI issues, and also the UI glitch you reported last time. Please review again :)

rossanafmenezes commented:
Thank! I will review it today 😄

rossanafmenezes reviewed (approved):
Looks good @toupper the UI issue is fixed. 🎉 

I m not merging in case @kidinov or @AnirudhBhat would like to take a second look.

kidinov commented:
@toupper @rossanafmenezes sorry, I won't have time today to look into that. If you feel confident, please go ahead and merge that

toupper commented:
> @toupper @rossanafmenezes sorry, I won't have time today to look into that. If you feel confident, please go ahead and merge that

Thanks for the reply! I will do that
